src/Microsoft.Extensions.WebSockets.Internal/WebSocketConnection.cs

```diff
@@ -249,8 +247,18 @@ private void WriteMaskingKey(Span<byte> buffer)
             _maskingKeyBuffer.CopyTo(buffer);
         }
 
+        /// <summary>
+        /// Terminates the socket abruptly.
+        /// </summary>
+        public void Abort()
+        {
+            _terminateReceiveCts.Cancel();
```

Comment:
Cancellation tokens are so messy. CancelPendingRead should really be enough. Just call CancelPendingRead when the token trips. Then change all of the logic to just look at IsCancelled.